### Task 1: Permutations and Combinations
Suppose we alter the Lady Tasting Tea experiment to involve twelve cups of tea. Six have the milk in first and the other six having tea in first. A person claims they have the special power of being able to tell whether the tea or the milk went into a cup first upon tasting it. You agree to accept their claim if they can tell which of the six cups in your experiment had the milk in first.

Calculate, using Python, the probability that they select the correct six cups. Here you should assume that they have no special powers in figuring it out, that they are just guessing. Remember to show and justify your workings in code and MarkDown cells.

Suppose, now, you are willing to accept one error. Once they select the six cups they think had the milk in first, you will give them the benefit of the doubt should they have selected at least five of the correct cups. Calculate the probability, assuming they have no special powers, that the person makes at most one error.

Would you accept two errors? Explain.

In [11]:
# Import dependencies ------
# Mathematical functions from the standard library.
# https://docs.python.org/3/library/math.html
import math
# Permutations and combinations.
# https://docs.python.org/3/library/itertools.html
import itertools
# Random selections.
# https://docs.python.org/3/library/random.html
import random
# Numerical structures and operations.
# https://numpy.org/doc/stable/reference/index.html#reference
import numpy as np
# Plotting.
# https://matplotlib.org/stable/contents.html
import matplotlib.pyplot as plt

# set variables ----------

# Number of cups of tea in total.
no_cups = 12 #k
# Number of cups of tea with milk in first.
no_cups_milk_first = 6 #n
# Number of cups of tea with tea in first.
no_cups_tea_first = 6 #n-k

The total number of possible combinations is calculated with combination formula (binomial coefficient )or the number of ways to choose 𝑘 items from a set of 𝑛 items, where order does not matter:
>$\binom{n}{k} = \frac{n!}{k!(n-k)!}$

- 𝑛! is the factorial of 𝑛, the total number of items.
- 𝑘! is the factorial of 𝑘, the number of items you're choosing.
- (𝑛−𝑘)! is the factorial of the remaining items.


In [12]:
# evaluation the choose formula (ref: lecture notes)
n,k = no_cups, no_cups_milk_first #assigning variables
choose = math.factorial(n)/(math.factorial(k)*(math.factorial(n-k)))
choose


924.0

In [13]:
# math.comb function does the same thing. (ref: lecture notes)
ways = math.comb(no_cups, no_cups_milk_first)
# Show.
ways

924

In [24]:
# we can show all possible combinations the 6 cups can be chosen
    # with list(itertools.combinations()function) 
# (ref: lecture notes)

# Create The cup labels.
labels = list(range(n)) #create a list containing integers from 0 to 7 to represent the cups
print("teacup labels: ", labels)
# Show the different ways (combinations) of selecting no_cups_milk_first out of no_cups cups of tea.
combinations = list(itertools.combinations(labels, k))
# Show.
combs=combinations
print("count combinations: ",len(combs))
combs

teacup labels:  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
count combinations:  924


[(0, 1, 2, 3, 4, 5),
 (0, 1, 2, 3, 4, 6),
 (0, 1, 2, 3, 4, 7),
 (0, 1, 2, 3, 4, 8),
 (0, 1, 2, 3, 4, 9),
 (0, 1, 2, 3, 4, 10),
 (0, 1, 2, 3, 4, 11),
 (0, 1, 2, 3, 5, 6),
 (0, 1, 2, 3, 5, 7),
 (0, 1, 2, 3, 5, 8),
 (0, 1, 2, 3, 5, 9),
 (0, 1, 2, 3, 5, 10),
 (0, 1, 2, 3, 5, 11),
 (0, 1, 2, 3, 6, 7),
 (0, 1, 2, 3, 6, 8),
 (0, 1, 2, 3, 6, 9),
 (0, 1, 2, 3, 6, 10),
 (0, 1, 2, 3, 6, 11),
 (0, 1, 2, 3, 7, 8),
 (0, 1, 2, 3, 7, 9),
 (0, 1, 2, 3, 7, 10),
 (0, 1, 2, 3, 7, 11),
 (0, 1, 2, 3, 8, 9),
 (0, 1, 2, 3, 8, 10),
 (0, 1, 2, 3, 8, 11),
 (0, 1, 2, 3, 9, 10),
 (0, 1, 2, 3, 9, 11),
 (0, 1, 2, 3, 10, 11),
 (0, 1, 2, 4, 5, 6),
 (0, 1, 2, 4, 5, 7),
 (0, 1, 2, 4, 5, 8),
 (0, 1, 2, 4, 5, 9),
 (0, 1, 2, 4, 5, 10),
 (0, 1, 2, 4, 5, 11),
 (0, 1, 2, 4, 6, 7),
 (0, 1, 2, 4, 6, 8),
 (0, 1, 2, 4, 6, 9),
 (0, 1, 2, 4, 6, 10),
 (0, 1, 2, 4, 6, 11),
 (0, 1, 2, 4, 7, 8),
 (0, 1, 2, 4, 7, 9),
 (0, 1, 2, 4, 7, 10),
 (0, 1, 2, 4, 7, 11),
 (0, 1, 2, 4, 8, 9),
 (0, 1, 2, 4, 8, 10),
 (0, 1, 2, 4, 8, 11),
 (0, 1, 2, 4

Now we have 3 different ways of proving that there are 924 possible combinations of chosing 6 items out of 12.
$$\binom{n}{k} = \frac{n!}{k!(n-k)!} $$
$$\binom{12}{6} = \frac{12!}{6!(12-6)!} = 924$$


##### *Probability*
In this case, choosing the 6 correct cups from 12 is a mutually exclusive scenario where the lady either gets it all right or not. There are 924 different ways to choose 6 cups from 12. Therefore, she has a 1 in 924 (or 0.11%) chance of succeeding. 
$$ P= \frac {1}{\binom {n}{k}} = \frac {1}{\binom {12}{6}} = \frac {1}{924} $$

###### ref: combination formula: https://en.wikipedia.org/wiki/Combination<br>ref: binomial coefficient: https://en.wikipedia.org/wiki/Binomial_coefficient <br> ref: calculate probability, mutually exclusive: https://www.wikihow.com/Calculate-Probability

In [34]:
# Percentage chance
P = round((1/choose)*100,2)
print(f"{P}%")

0.11%


Now let's see what is the probability of success if we allow 1 error so we accept both 5 correct and 6 correct cups as successful trial.

##### Using the combination formula

$$\binom{n}{k} = \frac{n!}{k!(n-k)!} $$ 
<br>
>Total number of combinations: $\binom{6}{12} = \frac{12!}{6!(12-6)!} =924$ <br>
>Total number of correct answers: $\binom{6}{6} = \frac{6!}{6!(6-6)!} =1$ <br>
>Total number of answers with 5 error: $\binom{6}{5} = \frac{6!}{5!(6-5)!} =6$ <br>
>Total number of answers with 1 correct: $\binom{6}{1} = \frac{6!}{1!(6-1)!} =6$ <br>
As there are 6 possible sets of the 5 correct cups and 6 possible cups for the 1 incorrect cup, these are multipled 6x6=36

In [69]:
# create function to calculate combinationsinations
def combinations(n,k):
    C= math.factorial(n)//(math.factorial(k)*(math.factorial(n-k)))
    return(C)

# number of possible combinationsinations using the combinationsination formula
tot=combinations(12,6)
# number of possible combinationsinations for correct answer
correct=combinations(6,6)
# number of possible combinationsinations with 1 error (5 correct of 6 out of 12)
correct5=combinations(6,5)
# number of possible combinationsinations with 1 error (1 incorrect of 6 out of 12)
error1=combinations(6,1)

print(tot, correct, correct5,error1)


924 1 6 6


In [ ]:
# Calculate possible ways of correctly chosing 5 or 6 out of 12


In [40]:
# Percentage chance
P = round((2/choose)*100,2)
print(f"{P}%")

0.22%
